core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java

```diff
@@ -0,0 +1,289 @@
+        for (int channel = 0; channel < pageChannels; channel++) {
+            blocksWithRowId[channel] = pageWithoutRowIdBlock.getBlock(channel);
+        }
+        blocksWithRowId[pageChannels] = makeOutputRowIdBlock(rowIdBlock, totalPositions, rowIdPositions);
+
+        return new Page(blocksWithRowId);
+    }
+
+    private static Block makeOutputRowIdBlock(Block rowIdBlock, int totalPositions, int[] rowIdPositions)
+    {
+        Block underlyingBlock = getUnderlyingBlock(rowIdBlock);
+        if (rowIdBlock.allPositionsAreNull()) {
+            return getAllNullsRowIdBlock(rowIdBlock, underlyingBlock, totalPositions);
+        }
+        else {
+            if (underlyingBlock instanceof RowBlock) {
+                List<Block> newRowIdChildrenBuilder = new ArrayList<>();
+                rowIdBlock.getChildren().stream()
+                        .map(block -> block.getPositions(rowIdPositions, 0, totalPositions))
+                        .forEach(newRowIdChildrenBuilder::add);
+                return RowBlock.fromFieldBlocks(
+                        totalPositions,
+                        Optional.empty(),
+                        newRowIdChildrenBuilder.toArray(new Block[] {}));
+            }
+            else {
+                return rowIdBlock.getPositions(rowIdPositions, 0, totalPositions);
+            }
+        }
+    }
+
+    private static Block getAllNullsRowIdBlock(Block rowIdBlock, Block underlyingBlock, int positionCount)
+    {
+        boolean[] nulls = new boolean[positionCount];
+        Arrays.fill(nulls, true);
+        if (underlyingBlock instanceof RowBlock) {
+            return RowBlock.fromFieldBlocks(positionCount, Optional.of(nulls), rowIdBlock.getChildren().toArray(new Block[]{}));
+        }
+        else {
+            return ArrayBlock.fromElementBlock(positionCount, Optional.of(nulls), new int[positionCount], underlyingBlock);
+        }
+    }
+
+    private void addDeleteRow(PageBuilder pageBuilder, Page originalPage, int position)
+    {
+        // Copy the write redistribution columns and the rowId column
+        for (int targetChannel : dataColumnChannels) {
+            int redistributionChannelNumber = redistributionChannelNumbers.get(targetChannel);
+            Type columnType = dataColumnTypes.get(targetChannel);
+            BlockBuilder targetBlock = pageBuilder.getBlockBuilder(targetChannel);
+
+            if (redistributionChannelNumber >= 0) {
+                // The value comes from that column of the page
+                columnType.appendTo(originalPage.getBlock(redistributionChannelNumber), position, targetBlock);
+            }
+            else {
+                // We don't care about the other data columns
+                targetBlock.appendNull();
+            }
+        }
+        // Add the operation column == deleted
+        pageBuilder.getBlockBuilder(dataColumnChannels.size()).writeInt(DELETE_OPERATION_NUMBER);
```

Comment:
All of the calls to `BlockBuilder.writeInt` and `Block.getInt` are incorrect, since blocks are owned by types, and thus all access should be through types:
```java
INTEGER.writeLong(builder, value);
```
```java
long value = INTEGER.getLong(block, position);
```
Only types should interact with blocks directly. This goes for the whole PR.

Reply:
Changed everywhere I could find.